common/src/main/java/io/pravega/common/io/filesystem/FileModificationEventWatcher.java

```diff
@@ -0,0 +1,225 @@
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * Watches for modifications to the specified file and performs the specified action (in the form of a callback) upon
+ * modification detection.
+ *
+ * This implementation uses NIO and event-based Java WatchService to watch for any modifications to the specified file.
+ *
+ * Note that:
+ *
+ * - The specified actions may trigger slightly later than when the file is actually modified.
+ *
+ * - If there are multiple file modifications in quick succession, only the last one may trigger the specified action.
+ *
+ * - The callback twice under rare circumstances when a file is modified. The Java {@link WatchService} raises two
+ *   events per modification: one for the content and second for the modification time. While there is a de-duplication
+ *   logic in this class to ensure that only a single event is handled and notified to the callback, it may not work
+ *   at times. If the calling code needs to be notified exactly once at all times, it should consider using
+ *   {@link FileModificationPollingMonitor} instead.
+ *
+ * - This class won't work for files that is under a symbolic link (pointing to a directory). This is because the
+ *   Java {@link WatchService} does not raise events for modifications of such files. Consider using
+ *   {@link FileModificationPollingMonitor} instead.
+ *
+ */
+@Slf4j
+public class FileModificationEventWatcher extends Thread implements FileModificationMonitor {
+
+    private static final AtomicInteger THREAD_NUM = new AtomicInteger();
+
+    /**
+     * The path of file to watch.
+     */
+    private final Path pathOfFileToWatch;
+
+    /**
+     * The action to perform when the specified file changes.
+     */
+    private final Consumer<WatchEvent<?>> callback;
+
+    private final UncaughtExceptionHandler uncaughtExceptionalHandler = (t, e) -> logException(e);
+
+    private boolean loopContinuously;
+
+    /**
+     * Creates a new instance.
+     *
+     * @param fileToWatch the file to watch
+     * @param callback    the callback to invoke when a modification to the {@code fileToWatch} is detected
+     * @throws NullPointerException if either {@code fileToWatch} or {@code callback} is null
+     * @throws InvalidPathException if {@code fileToWatch} is invalid
+     * @throws FileNotFoundException when a file at specified path {@code fileToWatch} does not exist
+     */
+    public FileModificationEventWatcher(@NonNull String fileToWatch, @NonNull Consumer<WatchEvent<?>> callback)
+            throws FileNotFoundException {
+        this(Paths.get(fileToWatch), callback, true);
+    }
+
+    /**
+     * Creates a new instance.
+     *
+     * @param fileToWatch path of the file to watch
+     * @param callback          the callback to invoke when a modification to the {@code fileToWatch} is detected
+     * @param loopContinuously  whether to keep the thread look for file modification after one iteration. This option
```

Comment:
"thread look"? Is it "thread lock"? If so please define this term as it's not standard.

Reply:
That was a typo. I've modified the comment now.